python/ray/data/tests/test_map.py

```diff
@@ -330,6 +331,126 @@ def map_generator(item: dict) -> Iterator[int]:
+
+    # Add 1ns to timestamp
+    row["timestamp"] = row["timestamp"] + pd.Timedelta(1, "ns")
+    return row
+
+
+# Helper function to create a sample timestamp nanosecs DataFrame
+def create_timestamp_dataframe():
+    return pd.DataFrame(
+        {
+            "id": [1, 2, 3],
+            "timestamp": pd.to_datetime(
+                [
+                    "2024-01-01 00:00:00.123456789",
+                    "2024-01-02 00:00:00.987654321",
+                    "2024-01-03 00:00:00.111222333",
+                ]
+            ),
+            "value": [10.123456789, 20.987654321, 30.111222333],
+        }
+    )
+
+
+# Handle Timestamp nanoseconds
+@pytest.mark.parametrize(
+    "df, expected_df",
+    [
+        (
+            create_timestamp_dataframe(),
+            pd.DataFrame(
+                {
+                    "id": [1, 2, 3],
+                    "timestamp": pd.to_datetime(
+                        [
+                            "2024-01-01 00:00:00.123456790",
+                            "2024-01-02 00:00:00.987654322",
+                            "2024-01-03 00:00:00.111222334",
+                        ]
+                    ),
+                    "value": [10.123456789, 20.987654321, 30.111222333],
+                }
+            ),
+        )
+    ],
+)
+def test_map_timestamp_nanosecs(df, expected_df, ray_start_regular_shared):
+    ray_data = ray.data.from_pandas(df)
+    result = ray_data.map(process_timestamp_data)
+    processed_df = result.to_pandas()
+    assert processed_df.shape == df.shape, "DataFrame shapes do not match"
+    pd.testing.assert_frame_equal(processed_df, expected_df)
+    assert (processed_df["timestamp"] - df["timestamp"]).iloc[0] == pd.Timedelta(
+        1, "ns"
+    )
+    assert (processed_df["timestamp"] - df["timestamp"]).iloc[1] == pd.Timedelta(
+        1, "ns"
+    )
+    assert (processed_df["timestamp"] - df["timestamp"]).iloc[2] == pd.Timedelta(
+        1, "ns"
+    )
+
+
+# Handle numpy.datetime64 values
+@pytest.mark.parametrize(
+    "df",
+    [
+        pd.DataFrame(
+            {
+                "id": [1, 2, 3],
+                "timestamp": [
+                    np.datetime64("2024-01-01T00:00:00.123456789"),
+                    np.datetime64("2024-01-02T00:00:00.987654321"),
+                    np.datetime64("2024-01-03T00:00:00.111222333"),
+                ],
+                "value": [10.123456789, 20.987654321, 30.111222333],
+            }
+        )
+    ],
+)
+def test_map_numpy_datetime(df, ray_start_regular_shared):
+    ray_data = ray.data.from_pandas(df)
+    result = ray_data.map(process_timestamp_data)
+    processed_df = result.to_pandas()
+    # Ensure numpy.datetime64 is correctly converted to pandas Timestamp
+    assert isinstance(processed_df["timestamp"].iloc[0], pd.Timestamp)
+
+    # Check that the timestamp has been incremented by 1ns
+    assert (processed_df["timestamp"] - df["timestamp"]).min() == pd.Timedelta(1, "ns")
```

Comment:
Can we standardize how we are checking the diff of 1ns? Seems like we are doing this, the equality check with the expected check and also individually checking the differences. Not super opinionated on which is the clearest but let's just use one (unless I am missing some subtlety here)
.

Reply:
This one is datetime vs np.datetime64. Let me consolidate this code better.